stan/math/rev/fun/rows_dot_product.hpp

```diff
@@ -17,12 +16,43 @@ template <typename Mat1, typename Mat2,
           require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
 inline Eigen::Matrix<var, Mat1::RowsAtCompileTime, 1> rows_dot_product(
     const Mat1& v1, const Mat2& v2) {
-  check_matching_sizes("dot_product", "v1", v1, "v2", v2);
-  Eigen::Matrix<var, Mat1::RowsAtCompileTime, 1> ret(v1.rows(), 1);
-  for (size_type j = 0; j < v1.rows(); ++j) {
-    ret.coeffRef(j) = dot_product(v1.row(j), v2.row(j));
+  check_matching_sizes("rows_dot_product", "v1", v1, "v2", v2);
+
+  ref_type_t<Mat1> v1_ref = v1;
+  ref_type_t<Mat2> v2_ref = v2;
+
+  arena_matrix<promote_scalar_t<double, Mat1>> arena_v1_val = value_of(v1_ref);
+  arena_matrix<promote_scalar_t<double, Mat2>> arena_v2_val = value_of(v2_ref);
+
+  arena_matrix<promote_scalar_t<var, Mat1>> arena_v1;
+  arena_matrix<promote_scalar_t<var, Mat2>> arena_v2;
+
+  if (!is_constant<Mat1>::value) {
+    arena_v1 = v1_ref;
+  }
+
+  if (!is_constant<Mat2>::value) {
+    arena_v2 = v2_ref;
   }
```

Comment:
How would yinz feel about something like 
```suggestion
  using arena_v1_type = conditional_arena_matrix<promote_scalar_t<var, Mat1>, !is_constant<Mat1>::value>;
  arena_v1_type arena_v1(v1_ref);
  using arena_v2_type = conditional_arena_matrix<promote_scalar_t<var, Mat2>, !is_constant<Mat2>::value>;
  arena_v2_type arena_v2(v2_ref);
```

So in the case of the RHS being `false` the `conditional_arena_matrix` is an `Eigen::Map<promote_scalar_t<var, Mat2>>(nullptr, 0, 0)`

Reply:
That is more or less how the new `to_arena_if` in https://github.com/stan-dev/math/pull/2026 works. After that is merged this can be:
```suggestion
  arena_matrix<promote_scalar_t<var, Mat1>> arena_v1 = to_arena_if<!is_constant<Mat1>::value>(v1_ref);
  arena_matrix<promote_scalar_t<var, Mat2>> arena_v2 = to_arena_if<!is_constant<Mat2>::value>(v2_ref);
```

Reply:
Nice! Yeah I was doing this on the other branch I was playing around in